atlassian_jwt_auth/contrib/server/helpers.py

```diff
@@ -1,45 +1,40 @@
-import logging
-
+from jwt.compat import text_type
 import jwt.exceptions
-import requests.exceptions
-
-from atlassian_jwt_auth.exceptions import (
-    PublicKeyRetrieverException,
-)
 
 
-def _requires_asap(verifier, auth, parse_jwt_func, response_class,
+def _requires_asap(get_verifier, auth, parse_jwt_func, build_response,
                    asap_claim_holder,
                    verify_issuers_func=None,
                    issuers=None,
                    ):
     """ Internal code used in various requires_asap decorators. """
-    message, exception = None, None
-    if not auth or len(auth) != 2:
-        return response_class('Unauthorized', status=401)
+    if isinstance(auth, text_type):
+        # Per PEP-3333, headers must be in ISO-8859-1 or use an RFC-2047
+        # MIME encoding. We don't really care about MIME encoded
+        # headers, but some libraries allow sending bytes (Django tests)
+        # and some (requests) always send str so we need to convert if
+        # that is the case to properly support Python 3.
+        auth = auth.encode(encoding='iso-8859-1')
+    try:
+        scheme, auth = auth.split(b' ')
+    except ValueError:
+        scheme = b''
+
+    if scheme.lower() != b'bearer':
+        return build_response('Unauthorized', status=401, headers={
+                              'WWW-Authenticate': 'Bearer'})
+    verifier = get_verifier()
     try:
-        asap_claims = parse_jwt_func(verifier, auth[1])
+        asap_claims = parse_jwt_func(verifier, auth)
         if verify_issuers_func is not None:
             verify_issuers_func(asap_claims, issuers)
         asap_claim_holder.asap_claims = asap_claims
-    except requests.exceptions.HTTPError as e:
```

Comment:
Actually this can occur if the key cannot be found and that's not the server's fault at all.

Reply:
Good point.

This should probably be handled at the point where the request is made: any 4xx response from the key server should raise an `InvalidTokenError` (eventually resulting in a 401), while any other error (can't reach remote server, 5xx response, etc.) should be a `PublicKeyRetrieverException`.

Maybe `InvalidIssuerError` could represent a 4xx response from the key server, and something like `ForbiddenIssuerException` could be a valid issuer that isn't in the whitelist (with a 403 response)?